local-cluster/tests/local_cluster.rs

```diff
@@ -6770,3 +6791,302 @@ fn test_alpenglow_ensure_liveness_after_double_notar_fallback() {
+            if *total_weight >= TOTAL_STAKE * 6 / 10 {
+                self.post_experiment_roots.insert(slot);
+            }
+        }
+
+        fn sufficient_roots_created(&self) -> bool {
+            self.post_experiment_roots.len() >= 10
+        }
+    }
+
+    // Start vote monitoring thread
+    let vote_listener_thread = std::thread::spawn({
+        let node_c_turbine_disabled = node_a_turbine_disabled.clone();
+
+        move || {
+            let mut buffer = [0u8; 65_535];
+            let mut experiment_state = ExperimentState::new();
+
+            let timer = std::time::Instant::now();
+
+            loop {
+                let bytes_received = vote_listener_socket
+                    .recv(&mut buffer)
+                    .expect("Failed to receive vote data");
+
+                let bls_message = bincode::deserialize::<BLSMessage>(&buffer[..bytes_received])
+                    .expect("Failed to deserialize BLS message");
+
+                let BLSMessage::Vote(vote_message) = bls_message else {
+                    continue;
+                };
+
+                let vote = &vote_message.vote;
+                let node_index = vote_message.rank as usize;
+
+                // Stage timeouts
+                let elapsed_time = timer.elapsed();
+
+                for stage in Stage::all() {
+                    if elapsed_time > stage.timeout() {
+                        panic!(
+                            "Timeout during {} stage. node_c_turbine_disabled: {:#?}. Latest vote: {:#?}. Experiment state: {:#?}",
+                            stage.name(),
+                            node_c_turbine_disabled.load(Ordering::Acquire),
+                            vote,
+                            experiment_state
+                        );
+                    }
+                }
+
+                // Stage 1: Wait for stability, then introduce partition at slot 50
+                if vote.slot() == 50 && !node_c_turbine_disabled.load(Ordering::Acquire) {
+                    node_c_turbine_disabled.store(true, Ordering::Release);
+                    experiment_state.stage = Stage::ObserveSkipFallbacks;
+                }
+
+                // Stage 2: Monitor for expected fallback vote patterns
+                if experiment_state.stage == Stage::ObserveSkipFallbacks {
+                    experiment_state.record_vote_bitmap(vote.slot(), node_index, vote);
+
+                    // Check if we've observed the expected pattern for 3 consecutive slots
+                    if experiment_state.matches_expected_pattern() {
+                        node_c_turbine_disabled.store(false, Ordering::Release);
+                        experiment_state.stage = Stage::ObserveLiveness;
+                    }
+                }
+
+                // Stage 3: Verify continued liveness after partition resolution
+                if experiment_state.stage == Stage::ObserveLiveness && vote.is_finalize() {
```

Comment:
In this stage we can also check finalization cert right? Then you don't need to record and check total stake for finalization votes yourself. Just count how many new finalization certs you have observed.